frame/vamm/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+//!
+//! TODO
+//!
+//! ### Goals
+//!
+//! TODO
+//!
+//! ### Actors
+//!
+//! TODO
+//!
+//! (clearing house)
+//!
+//! ### Implementations
+//!
+//! The VAMM pallet provides implementations for the following traits:
+//!
+//! - [`Vamm`](composable_traits::vamm::Vamm): Exposes functionality for
+//! creating, managing and deprecating virtual automated market makers.
+//!
+//! ## Interface
+//!
+//! TODO
+//!
+//! ### Extrinsics
+//!
+//! TODO
+//!
+//! not applicable
+//!
+//! ### Implemented Functions
+//!
+//! * [`create`](<Pallet as Vamm>::create)
+//!
+//! TODO
+//!
+//! ### Public Functions
+//!
+//! TODO
+//!
+//! ### Public Storage Objects
+//!
+//! TODO
+//!
+//! ## Usage
+//!
+//! TODO
+//!
+//! ### Example
+//!
+//! TODO
+//!
+//! ## Related Modules
+//!
+//! TODO
+//!
+//! (clearing house)
+//!
+//! <!-- Original author: @Cardosaum -->
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(test)]
+mod mock;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	// ----------------------------------------------------------------------------------------------------
+	//                                       Imports and Dependencies
+	// ----------------------------------------------------------------------------------------------------
+
+	use codec::{Codec, FullCodec};
+	use composable_traits::vamm::*;
+	use frame_support::{pallet_prelude::*, sp_std::fmt::Debug, transactional, Blake2_128Concat};
+	use sp_runtime::{
+		traits::{AtLeast32BitUnsigned, CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, One, Zero},
+		ArithmeticError,
+	};
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                    Declaration Of The Pallet Type
+	// ----------------------------------------------------------------------------------------------------
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                             Config Trait
+	// ----------------------------------------------------------------------------------------------------
+
+	// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Event type emitted by this pallet. Depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The `VammId` used by the pallet. Corresponds to the Ids used by the Vamm pallet.
+		type VammId: Default
+			+ CheckedAdd
+			+ Clone
+			+ Copy
+			+ Debug
+			+ FullCodec
+			+ MaxEncodedLen
+			+ MaybeSerializeDeserialize
+			+ One
+			+ Parameter
+			+ PartialEq
+			+ TypeInfo
+			+ Zero;
+
+		/// Timestamp to be used for twap calculations and market deprecation.
```

Comment:
I've been thinking about using the terms 'stopping' or 'freezing' a market. Thoughts?

Reply:
Don't have a strong position in favor/against any of those.
Which one do you think is better, and why?

Reply:
Can markets be "continued" or "unfrozen" after they are stoped/frozen?

Reply:
I'd imagine that they are permanently stopped/frozen. Might have to discuss this with the rest of the team

Reply:
TBH, I think we could use 'closing a market' as in TradFi

Reply:
Alright, in a recent commit I proceeded with the change: 'Deprecation' -> 'Closing'